[Pallas GPU] `jnp.abs()` gives the wrong result

### Description

```python
import functools
import jax
from jax.experimental import pallas as pl
import jax.numpy as jnp
import numpy as np

def main():
    @jax.jit
    @functools.partial(
        pl.pallas_call,
        out_shape=jax.ShapeDtypeStruct((4, 4), jnp.float32),
    )
    def kernel(x_ref, o_ref):
        o_ref[...] = jnp.abs(x_ref[...])

    val = -3.2
    x = jnp.full((4, 4), val)
    out = kernel(x)
    expected = jnp.abs(x)
    np.testing.assert_array_equal(out, expected)

if __name__ == '__main__':
    main()
```

Expected behaviour:

No error.

Actual behaviour:

```
2024-08-22 13:22:04.157537: W external/xla/xla/service/gpu/nvptx_compiler.cc:836] The NVIDIA driver's CUDA version is 12.5 which is older than the PTX compiler version (12.6.20). Because the driver is older than the PTX compiler version, XLA is disabling parallel compilation, which may slow down compilation. You should update your NVIDIA driver or use the NVIDIA-provided CUDA forward compatibility packages.
Traceback (most recent call last):
  File "/home/ayx/test/main.py", line 23, in <module>
    main()
  File "/home/ayx/test/main.py", line 20, in main
    np.testing.assert_array_equal(out, expected)
  File "/home/ayx/test/venv/lib/python3.12/site-packages/numpy/_utils/__init__.py", line 85, in wrapper
    return fun(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/ayx/test/venv/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 1025, in assert_array_equal
    assert_array_compare(operator.__eq__, actual, desired, err_msg=err_msg,
  File "/home/ayx/miniconda3/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/ayx/test/venv/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 889, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

Mismatched elements: 16 / 16 (100%)
Max absolute difference among violations: 3.2
Max relative difference among violations: 1.
 ACTUAL: array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)
 DESIRED: array([[3.2, 3.2, 3.2, 3.2],
       [3.2, 3.2, 3.2, 3.2],
       [3.2, 3.2, 3.2, 3.2],
       [3.2, 3.2, 3.2, 3.2]], dtype=float32)
```

I noticed that there would be no error if I change the line

```python
x = jnp.full((4, 4), val)
```

to

```python
x = jnp.full((4, 4), val, dtype=jnp.float32)
```

In other words, the error only happens when `weak_type` is True.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.32.dev20240822+b56ed8eed
jaxlib: 0.4.31
numpy:  2.1.0
python: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:12:24) [GCC 11.2.0]
jax.devices (1 total, 1 local): [CudaDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='ayx-gpu-test-1', release='5.15.0-1062-gcp', version='#70~20.04.1-Ubuntu SMP Fri May 24 20:12:18 UTC 2024', machine='x86_64')


$ nvidia-smi
Thu Aug 22 13:26:34 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 555.52.04              Driver Version: 555.52.04      CUDA Version: 12.5     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-SXM4-40GB          Off |   00000000:00:04.0 Off |                    0 |
| N/A   32C    P0             59W /  400W |     526MiB /  40960MiB |      3%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A     37173      G   /usr/lib/xorg/Xorg                             65MiB |
|    0   N/A  N/A     37257      G   /usr/bin/gnome-shell                           17MiB |
|    0   N/A  N/A     37977      C   python                                        416MiB |
+-----------------------------------------------------------------------------------------+
```

--- a/jax/_src/pallas/triton/lowering.py
+++ b/jax/_src/pallas/triton/lowering.py
@@ -564,7 +564,7 @@ class _Extern:
     if len(avals) != len(self.arg_types):
       return False
     return all(
-        aval.weak_type or aval.dtype.name == arg_type
+        aval.dtype.name == arg_type
         for aval, arg_type in zip(avals, self.arg_types)
     )
 

--- a/tests/pallas/ops_test.py
+++ b/tests/pallas/ops_test.py
@@ -739,6 +739,17 @@ class OpsExtraTest(PallasBaseTest):
       x = jnp.array([0.42, 2.4]).astype(dtype)
       np.testing.assert_allclose(kernel(x), fn(x), rtol=1e-6)
 
+  def test_abs_weak_type(self):
+    # see https://github.com/google/jax/issues/23191
+    @functools.partial(
+        self.pallas_call, out_shape=jax.ShapeDtypeStruct((4, 4), jnp.float32),
+    )
+    def kernel(x_ref, o_ref):
+      o_ref[...] = jnp.abs(x_ref[...])
+
+    x = jnp.broadcast_to(-3.2, (4, 4))  # sets `weak_type` to `True`
+    np.testing.assert_allclose(kernel(x), jnp.abs(x), rtol=1e-6)
+
   @parameterized.parameters(
       ("float32", "int32"),
       ("float64", "int32"),

